src/lib/TopLevel.hs

```diff
@@ -775,18 +775,44 @@ evalLLVM block = do
   benchRequired <- requiresBench <$> getPassCtx
   nativeFun <- liftIO $ linkFunObjCode obj dyvarStores
     $ LinktimeVals reqFunPtrs reqDataPtrs
-  resultVals <-
-    liftIO $ callNativeFun nativeFun benchRequired logger [] resultTypes
-  resultValsNoPtrs <- mapM litValToPointerlessAtom resultVals
-  -- TODO: this unsafe coerce won't be needed once we treat everything as a
-  -- function, because then we won't need the reconstruction at all. If we were
-  -- going to keep it this way, the right thing to do would be to traverse the
-  -- atom and throw an error if we find any `BoxedRef`
-  Abs bs finalResult <- return reconAtom
-  let substVals = map SubstVal resultValsNoPtrs :: [SubstVal ImpNameC (Atom SimpToImpIR) ImpNameC n]
-  unsafeCoerceIRE <$> applySubst (bs @@> substVals) finalResult
+  dest@(TopDest _ destPtrs) <- makeDestTopLevel destTy
+  liftIO $ callNativeFun nativeFun benchRequired logger (PtrLit <$> destPtrs)
+  loadDestTopLevel dest
+
 {-# SCC evalLLVM #-}
 
+data TopDest (n::S) = TopDest (Atom SimpToImpIR n) [PtrLitVal]
+
+makeDestTopLevel :: (Mut n, Topper m) => SType n -> m n (TopDest n)
+makeDestTopLevel (RawRefTy valTy) = do
```

Comment:
Can we ensure that this `valTy` is simple? I'm a little afraid of leaking `makeDest` all the way up to here, since the plan is to make `Imp` a lot dumber than it is today. It shouldn't be able to handle most of the types we throw at it today